File: app/src/main/java/com/bureng/robocoinx/activity/SplashActivity.java
package com.bureng.robocoinx.activity;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.webkit.JavascriptInterface;
import android.webkit.WebView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;

import com.bureng.robocoinx.R;
import com.bureng.robocoinx.contract.SplashContract;
import com.bureng.robocoinx.presenter.SplashPresenter;
import com.bureng.robocoinx.utils.CacheContext;
import com.bureng.robocoinx.utils.FileManager;
import com.bureng.robocoinx.model.db.Fingerprint;
import com.bureng.robocoinx.model.request.SignUpRequest;
import com.bureng.robocoinx.model.view.ProfileView;
import com.bureng.robocoinx.utils.StaticValues;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

import java.io.InputStream;

public class SplashActivity extends Activity implements SplashContract.View {

    private SplashContract.Presenter presenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.splash_main);
        presenter = new SplashPresenter(this, this);
        if (ActivityCompat.checkSelfPermission(getApplicationContext(), android.Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED)
            ActivityCompat.requestPermissions(this, new String[]{android.Manifest.permission.WRITE_EXTERNAL_STORAGE}, 10);
        new Content().execute((Void)null);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == 10) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                //do here
                FileManager.getInstance().appendLog("first....");
            }
        }
    }

    @Override
    public void goHome(ProfileView profileView) {
        Intent intent = new Intent(getBaseContext(), HomeActivity.class);
        intent.putExtra(StaticValues.PROFILE_VIEW, profileView);
        startActivity(intent);
    }

    @Override
    public void showMessage(String message) {
        Toast.makeText(getApplicationContext(), message, Toast.LENGTH_SHORT).show();
    }

    @SuppressLint("SetJavaScriptEnabled")
    @Override
    public void loadFingerprint(SignUpRequest signupRequest) {
//        Fingerprint fingerprint = new Fingerprint("3269417398", "458b9e977a393c3d9897392a768db437");
//        new CacheContext<>(Fingerprint.class, getApplicationContext()).save(fingerprint, StaticValues.FINGERPRINT);
//        Intent intentlogin = new Intent(getApplicationContext(), LoginActivity.class);
//        intentlogin.putExtra(StaticValues.SIGNUP_REQ, signupRequest);
//        startActivity(intentlogin);

        String html;
        try {
            InputStream inputStream = getAssets().open("finger.html");
            Document document = Jsoup.parse(inputStream, "UTF-8", "");
            String doc = document.html();
            html = doc.replace("[code]", signupRequest.script);
            WebView webView = findViewById(R.id.webviewx);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                webView.getSettings().setSafeBrowsingEnabled(false);
            }
            webView.getSettings().setJavaScriptEnabled(true);
            webView.addJavascriptInterface(new WebAppInterface(this, signupRequest), "Android");
            webView.loadDataWithBaseURL("blarg://ignored", html, "text/html", "UTF-8", "");
        }catch (Exception e){
            e.printStackTrace();
            FileManager.getInstance().appendLog(e);
        }
    }

    public class WebAppInterface {
        SignUpRequest signupRequest;
        Context context;

        WebAppInterface(Context c, SignUpRequest s) {
            context = c;
            signupRequest = s;
        }

        @JavascriptInterface
        public void getFingerprint(String data) {
            if(!data.contains("error")){
                String[] split = data.split(":");
                Fingerprint fingerprint = new Fingerprint(split[0], split[1]);
                new CacheContext<>(Fingerprint.class, context).save(fingerprint, StaticValues.FINGERPRINT);
                goToLogin(context, signupRequest);
            }
        }

        private void goToLogin(Context context, SignUpRequest signupRequest) {
            Intent intent = new Intent(context, LoginActivity.class);
            intent.putExtra(StaticValues.SIGNUP_REQ, signupRequest);
            context.startActivity(intent);
        }
    }

    @SuppressLint("StaticFieldLeak")
    public class Content extends AsyncTask<Void, Void, Object> {

        @Override
        protected Object doInBackground(Void... voids) {
            try {
                return presenter.getResponse();
            }catch (Exception e){
                e.printStackTrace();
                FileManager.getInstance().appendLog(e);
            }
            return null;
        }

        @Override
        protected void onPostExecute(Object obj) {
            if (obj instanceof ProfileView)
                goHome((ProfileView)obj);
            else if(obj instanceof SignUpRequest)
                loadFingerprint((SignUpRequest)obj);
            else
                showMessage((String)obj);
        }
    }
}

File: app/src/main/java/com/bureng/robocoinx/activity/SplashActivity.kt
package com.bureng.robocoinx.activity

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import android.webkit.JavascriptInterface
import android.webkit.WebView
import android.widget.Toast
import androidx.core.app.ActivityCompat
import com.bureng.robocoinx.R
import com.bureng.robocoinx.contract.SplashContract
import com.bureng.robocoinx.model.common.DoAsync
import com.bureng.robocoinx.model.db.Fingerprint
import com.bureng.robocoinx.model.request.SignUpRequest
import com.bureng.robocoinx.model.view.ProfileView
import com.bureng.robocoinx.presenter.SplashPresenter
import com.bureng.robocoinx.utils.CacheContext
import com.bureng.robocoinx.utils.FileManager
import com.bureng.robocoinx.utils.StaticValues
import kotlinx.android.synthetic.main.splash_main.*
import org.jsoup.Jsoup

class SplashActivity : Activity(), SplashContract.View {
    private lateinit var presenter: SplashContract.Presenter
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.splash_main)
        if (ActivityCompat.checkSelfPermission(applicationContext, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 10)
        presenter = SplashPresenter(this, this)
        DoAsync{
            presenter.loadActivity()
        }.execute()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 10) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                FileManager.getInstance().appendLog("first....")
            }
        }
    }

    override fun goHome(profileView: ProfileView) {
        val intent = Intent(baseContext, HomeActivity::class.java)
        intent.putExtra(StaticValues.PROFILE_VIEW, profileView)
        startActivity(intent)
    }

    override fun showMessage(message: String?) {
        runOnUiThread { Toast.makeText(applicationContext, message, Toast.LENGTH_SHORT).show()}
    }

    @SuppressLint("SetJavaScriptEnabled")
    override fun loadFingerprint(signUpRequest: SignUpRequest) {
        val html: String
        try {
            val inputStream = assets.open("finger.html")
            val document = Jsoup.parse(inputStream, "UTF-8", "")
            val doc = document.html()
            html = doc.replace("[code]", signUpRequest.script)
            runOnUiThread {
                val webView = findViewById<WebView>(R.id.webviewx)
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    webView.settings.safeBrowsingEnabled = false
                }
                webView.settings.javaScriptEnabled = true
                webView.addJavascriptInterface(WebAppInterface(this, signUpRequest), "Android")
                webView.loadDataWithBaseURL("blarg://ignored", html, "text/html", "UTF-8", "")
            }
        } catch (e: Exception) {
            e.printStackTrace()
            FileManager.getInstance().appendLog(e)
        }
    }

    override fun showProgressBar() {
        runOnUiThread {
            progressBarSP.visibility = View.VISIBLE
            window.setFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE,
                    WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
        }
    }

    override fun hideProgressBar() {
        runOnUiThread {
            progressBarSP.visibility = View.GONE
            window.clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
        }
    }

    inner class WebAppInterface internal constructor(var context: Context, private var signUpRequest: SignUpRequest) {
        @JavascriptInterface
        fun getFingerprint(data: String) {
            if (!data.contains("error")) {
                val split = data.split(":".toRegex()).toTypedArray()
                val fingerprint = Fingerprint(split[0], split[1])
                CacheContext(Fingerprint::class.java, context).save(fingerprint, StaticValues.FINGERPRINT)
                goToLogin(context, signUpRequest)
            }
        }

        private fun goToLogin(context: Context, signUpRequest: SignUpRequest) {
            val intent = Intent(context, SignUpActivity::class.java)
            intent.putExtra(StaticValues.SIGNUP_REQ, signUpRequest)
            context.startActivity(intent)
        }
    }
}